File: tests/ut/cpp/ops/test_ops_relu.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <vector>
#include <memory>
#include "common/common_test.h"
#include "ops/ops_func_impl/relu.h"
#include "ops/auto_generate/gen_ops_name.h"
#include "abstract/abstract_value.h"
#include "ops/test_ops.h"
#include "ops/test_ops_dyn_cases.h"
#include "ops/test_ops_cmp_utils.h"

namespace mindspore {
namespace ops {
class TestReLU : public TestOps,
                 public testing::WithParamInterface<std::tuple<EltwiseOpShapeParams, EltwiseOpTypeParams>> {};

TEST_P(TestReLU, dyn_shape) {
  const auto &shape_param = std::get<0>(GetParam());
  const auto &dtype_param = std::get<1>(GetParam());
  auto x = std::make_shared<abstract::AbstractTensor>(dtype_param.x_type, shape_param.x_shape);
  ASSERT_NE(x, nullptr);
  auto expect_shape = std::make_shared<abstract::Shape>(shape_param.out_shape);
  auto expect_type = std::make_shared<TensorType>(dtype_param.out_type);
  DoFuncImplInferAndCompare<ReLUFuncImpl>(kNameReLU, {x}, expect_shape, expect_type);
}

namespace {
auto ReLUOpTypeCases = testing::ValuesIn({
  EltwiseOpTypeParams{kInt8, kInt8},
  EltwiseOpTypeParams{kInt16, kInt16},
  EltwiseOpTypeParams{kInt32, kInt32},
  EltwiseOpTypeParams{kInt64, kInt64},
  EltwiseOpTypeParams{kUInt8, kUInt8},
  EltwiseOpTypeParams{kUInt16, kUInt16},
  EltwiseOpTypeParams{kUInt32, kUInt32},
  EltwiseOpTypeParams{kUInt64, kUInt64},
  EltwiseOpTypeParams{kFloat16, kFloat16},
  EltwiseOpTypeParams{kFloat32, kFloat32},
  EltwiseOpTypeParams{kFloat64, kFloat64},
});
}

INSTANTIATE_TEST_CASE_P(TestReLUGroup, TestReLU, testing::Combine(EltwiseDynShapeTestCases, ReLUOpTypeCases));
}  // namespace ops
}  // namespace mindspore

File: tests/st/ops/test_ops_relu.py
# Copyright 2023 Huawei Technoreluies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
from tests.mark_utils import arg_mark
import pytest
import numpy as np
import mindspore as ms
from mindspore import context, Tensor
from mindspore.ops import relu
from tests.st.utils import test_utils
from tests.st.ops.dynamic_shape.test_op_utils import TEST_OP

def generate_random_input(shape, dtype):
    return np.random.randn(*shape).astype(dtype)

def generate_expect_forward_output(x, dtype):
    return np.maximum(x, 0).astype(dtype)

def generate_expect_backward_output(x, dtype):
    return np.where(x > 0, 1, 0).astype(dtype)

@test_utils.run_with_cell
def relu_forward_func(x):
    return relu(x)

@test_utils.run_with_cell
def relu_backward_func(x):
    return ms.ops.grad(relu_forward_func, (0))(x)

@test_utils.run_with_cell
def relu_vmap_func(x):
    return ms.ops.vmap(relu_forward_func, in_axes=0, out_axes=0)(x)


@arg_mark(plat_marks=['platform_ascend', 'platform_gpu', 'cpu_linux', 'cpu_windows', 'cpu_macos'], level_mark='level0', card_mark='onecard', essential_mark='essential')
@pytest.mark.parametrize("mode", [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
def test_relu_normal(mode):
    """
    Feature: test relu operator
    Description: test relu run by pyboost
    Expectation: success
    """
    context.set_context(mode=mode)

    ## forward
    np_array = np.random.rand(2, 3, 4)
    x = Tensor(np_array, ms.float32)
    output = relu_forward_func(x)
    expect = generate_expect_forward_output(np_array, np.float32)
    assert np.allclose(output.asnumpy(), expect)

    ## backward
    np_array = np.random.rand(2, 3, 4).astype(np.float32)
    x = Tensor(np_array, ms.float32)
    output = relu_backward_func(x)
    expect = generate_expect_backward_output(np_array, np.float32)
    assert np.allclose(output.asnumpy(), expect)


@arg_mark(plat_marks=['platform_ascend', 'platform_gpu', 'cpu_linux', 'cpu_windows', 'cpu_macos'], level_mark='level1', card_mark='onecard', essential_mark='unessential')
@pytest.mark.parametrize("mode", [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
def test_ops_relu_vmap(mode):
    """
    Feature: pyboost function.
    Description: test function relu vmap feature.
    Expectation: expect correct result.
    """
    context.set_context(mode=mode)
    x = generate_random_input((2, 3, 4, 5), np.float32)
    output = relu_vmap_func(Tensor(x))
    expect = generate_expect_forward_output(x, np.float32)
    np.testing.assert_allclose(output.asnumpy(), expect)


@arg_mark(plat_marks=['platform_ascend910b'], level_mark='level1', card_mark='onecard', essential_mark='unessential')
@pytest.mark.parametrize("mode", [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
def test_relu_bfloat16(mode):
    """
    Feature: test relu operator
    Description: test relu run by pyboost
    Expectation: success
    """
    context.set_context(mode=mode)
    np_array = np.random.rand(2, 3, 4)
    x = Tensor(np_array, ms.bfloat16)
    output = relu_forward_func(x)
    expect = generate_expect_forward_output(np_array, np.float32)
    assert np.allclose(output.float().asnumpy(), expect, rtol=2e-3, atol=2e-3)


@arg_mark(plat_marks=['platform_ascend', 'platform_gpu', 'cpu_linux', 'cpu_windows', 'cpu_macos'], level_mark='level1', card_mark='onecard', essential_mark='unessential')
def test_relu_dynamic():
    """
    Feature: test dynamic by TEST_OP.
    Description: test ops.relu dynamic shape feature.
    Expectation: expect correct result.
    """
    input_case1 = Tensor(np.random.rand(3, 4, 5, 6).astype(np.float32))
    input_case2 = Tensor(np.random.rand(3, 4).astype(np.float32))
    TEST_OP(relu_forward_func, [[input_case1], [input_case2]], 'relu')
